File: eval/cubeTest.py
# -*- encoding: utf-8 -*-
from collections import OrderedDict
import sys
import math

MAX_JUDGMENT = 4
MAX_HEIGHT = 5
beta = 1
gamma = 0.5

# $topic $docno $subtopic $judgement
qrels = {}
#$topic $subtopic $area
subtopic_weight = {}
# $topic $subtopic $gainHeights
current_gain_height = {}
# $topic $subtopic $occurrences
subtopic_cover = {}
# $docID $docLength
docLengthMap = {}

#########################################

#### Read qrels file(groundtruth), check format, and sort
def prepare_qrels(qrelsfile):
    tmp_qrels = {}
    count = 0
    for line in open(qrelsfile):
        topic, subtopic, docno, passage, judgment = line.strip().split('\t')
        if int(judgment) > 0 :
            judgment = int(judgment)
        else :
            judgment = 1
        if not tmp_qrels.has_key(topic):
            tmp_qrels[topic] = {}
            tmp_qrels[topic][docno] = {}
            tmp_qrels[topic][docno][subtopic] = {}
        if not tmp_qrels[topic].has_key(docno):
            tmp_qrels[topic][docno] = {}
            tmp_qrels[topic][docno][subtopic] = {}
        if not tmp_qrels[topic][docno].has_key(subtopic):
            tmp_qrels[topic][docno][subtopic] = {}

        tmp_qrels[topic][docno][subtopic][passage] = judgment

    for topic, docs in tmp_qrels.iteritems():
        for docno, subtopics in docs.iteritems():
            for subtopic, rels in subtopics.iteritems():
                rels = sorted(rels.values(),reverse=True)  
                log2 = math.log(2)
                rel = sum([ rel/(math.log(rank+2)/log2)  for rank, rel in enumerate(rels)])
                
                if not qrels.has_key(topic):
                    qrels[topic] = {}
                    subtopic_weight[topic] = {}
                    current_gain_height[topic] = {}
                    subtopic_cover[topic] = {}
                    qrels[topic][docno] = {}
                    qrels[topic][docno][subtopic] = {}
                if not qrels[topic].has_key(docno):
                    qrels[topic][docno] = {}
                    qrels[topic][docno][subtopic] = {}
                if not qrels[topic][docno].has_key(subtopic):
                    qrels[topic][docno][subtopic] = {}
                
                qrels[topic][docno][subtopic] = rel
                subtopic_weight[topic][subtopic] = 1
                current_gain_height[topic][subtopic] = 0
                subtopic_cover[topic][subtopic] = 0

    #### Normalize subtopic weight

    for topic, subtopics in subtopic_weight.iteritems():
        max_weight = get_max_weight(topic)
        for subtopic in subtopics:
            subtopic_weight[topic][subtopic] /= float(max_weight)

def get_doc_gain(topic,docno):
    gain = 0
    for subtopic, area in subtopic_weight[topic].iteritems():
        nrel = subtopic_cover[topic][subtopic]
        if qrels[topic].has_key(docno):
            if not qrels[topic][docno].has_key(subtopic):
                continue
        else:
            continue
        hight_keepfilling = get_hight_keepfilling(topic, docno, subtopic, nrel+1)
        area = get_area(topic,subtopic)
        
        gain += area*hight_keepfilling
        subtopic_cover[topic][subtopic] +=1
    return gain

def get_hight_keepfilling(topic, docno, subtopic, nrel):
    
    rel = 0
    if qrels[topic].has_key(docno):

        if qrels[topic][docno].has_key(subtopic):
            rel = qrels[topic][docno][subtopic]
    if rel == 0:
        return 0
    
    current_gain = current_gain_height[topic][subtopic]

    gain = get_hight_discount(nrel)*rel
    if current_gain + gain > MAX_HEIGHT: 
        gain = MAX_HEIGHT - current_gain
    current_gain_height[topic][subtopic] += gain
    return gain

def get_area(topic, subtopic):
    if subtopic_weight[topic].has_key(subtopic):
        return subtopic_weight[topic][subtopic]
    return 0

def get_hight_discount(nrels):
    return gamma ** nrels

def get_max_weight(topic):
    max_weight = sum([v for v in subtopic_weight[topic].values()])
    return max_weight


def evaluate_run(runfile,iterations):
    #### Process runs: compute measures for each topic and average
    results = OrderedDict()
    for line in open(runfile):
        topic, iteration, docno, score, ontopic, subtopics = line.split("\t")
        iteration = int(iteration)
        if not results.has_key(topic):
            results[topic] = OrderedDict()
            results[topic][iteration] = []
        elif not results[topic].has_key(iteration):
            results[topic][iteration] = []
        results[topic][iteration].append(docno)

    all_ct = {}
    all_act = {}
    for i in range(iterations):
        all_ct[i] = 0
        all_act[i] = 0

    ntopics = 0
    for topic, result in results.iteritems():
        if not qrels.has_key(topic):
            continue
        ntopics += 1
        time = 0
        score = 0
        ct_accu = 0
        for iteration in range(iterations):
            if result.has_key(iteration):
                for docno in result[iteration]:
                    gain =  get_doc_gain(topic,docno)
                    score += gain
                    act = score/MAX_HEIGHT
                    ct_accu += act/(iteration+1)
                    time += 1
                    
                ct = score/MAX_HEIGHT
                ct_speed = ct/(iteration+1)
                all_ct[iteration] += ct_speed
                all_act[iteration] += ct_accu/time
    cts = []
    acts = []
    for i in range(iterations):
        cts.append(all_ct[i]/ntopics)
        acts.append(all_act[i]/ntopics)
    return cts, acts

def evaluate_run_q(runfile,iterations):
    #### Process runs: compute measures for each topic and average
    results = OrderedDict()
    for line in open(runfile):
        topic, iteration, docno, score, ontopic, subtopics = line.split("\t")
        iteration = int(iteration)
        if not results.has_key(topic):
            results[topic] = OrderedDict()
            results[topic][iteration] = []
        elif not results[topic].has_key(iteration):
            results[topic][iteration] = []
        results[topic][iteration].append(docno)

    all_ct = {}
    all_act = {}
    for i in range(iterations):
        all_ct[i] = 0
        all_act[i] = 0

    ntopics = 0
    cts = {}
    acts = {}
    for topic, result in results.iteritems():
        if not qrels.has_key(topic):
            continue
        ntopics += 1
        time = 0
        score = 0
        ct_accu = 0
        for iteration in range(iterations):
            if result.has_key(iteration):
                for docno in result[iteration]:
                    gain =  get_doc_gain(topic,docno)
                    score += gain
                    act = score/MAX_HEIGHT
                    ct_accu += act/(iteration+1)
                    time += 1
                    
                ct = score/MAX_HEIGHT
                ct_speed = ct/(iteration+1)
                all_ct[iteration] += ct_speed
                all_act[iteration] += ct_accu/time

        cts[topic] = ct_speed
        acts[topic] = ct_accu/time

    return cts, acts

def main():
    if len(sys.argv) < 3:
        print "Usage: python cubeTest.py qrels runfile iterations"

    qrelsfile = sys.argv[1]
    runfile = sys.argv[2]
    iterations = int(sys.argv[3])

    prepare_qrels(qrelsfile)

    #### Process runs: compute measures for each topic and average
    results = OrderedDict()
    for line in open(runfile):
        topic, iteration, docno, score, ontopic, subtopics = line.split("\t")
        iteration = int(iteration)
        if not results.has_key(topic):
            results[topic] = OrderedDict()
            results[topic][iteration] = []
        elif not results[topic].has_key(iteration):
            results[topic][iteration] = []
        results[topic][iteration].append(docno)

    all_ct = {}
    all_act = {}
    for i in range(iterations):
        all_ct[i] = 0
        all_act[i] = 0


    print "run_id,topic,ct@10,avg_ct@10"
    ntopics = 0
    for topic, result in results.iteritems():
        if not qrels.has_key(topic):
            continue
        ntopics += 1
        time = 0
        score = 0
        ct_accu = 0
        for iteration in range(iterations):
            if result.has_key(iteration):
                for docno in result[iteration]:
                    gain =  get_doc_gain(topic,docno)
                    score += gain
                    act = score/MAX_HEIGHT
                    ct_accu += act/(iteration+1)
                    time += 1
                    
                ct = score/MAX_HEIGHT
                ct_speed = ct/(iteration+1)
                all_ct[iteration] += ct_speed
                all_act[iteration] += ct_accu/time
                print "ct@%d\t%s\t%.10f" % (iteration+1, topic, ct_speed)
                print "avg_ct@%d\t%s\t%.10f" % (iteration+1, topic, ct_accu/time)

        
    for i in range(iterations):
        print "ct@%d\tall\t%.10f" % ( i+1, all_ct[i]/ntopics)
        print "avg_ct@%d\tall\t%.10f" % (i+1, all_act[i]/ntopics)



if __name__ == "__main__":
    main()



    
   

File: core/src/main/java/br/ufmg/dcc/latin/metrics/CubeTest.java
package br.ufmg.dcc.latin.metrics;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

public class CubeTest {
	
	private int MAX_HEIGHT = 5;
	private double gamma = 0.5f;
	
	// $topic $docno $subtopic $judgement
	private Map<String, Map<String, Map<String, Double> > >  qrels;
	// $topic $subtopic $area
	private Map<String, Map<String, Double > >  subtopicWeight;
	// subtopic gain
	private Map<String, Double> currentGainHeight;
	// subtopic cover
	private Map<String, Integer> subtopicCover;
	
	public CubeTest(){
		qrels = new HashMap<String, Map<String, Map<String, Double> > > ();
		subtopicWeight = new HashMap<String, Map<String, Double > >();

		Map<String, Map<String, Map<String, List<Integer>> > >   tmpQrels = new HashMap<String, Map<String, Map<String, List<Integer> > > > ();
		try (BufferedReader br = new BufferedReader(new FileReader("../share/dd_qrels.txt"))) {
			String line;
			while ((line = br.readLine()) != null) {
				String[] splitedLine = line.split("\t",5);
				
				String topic = splitedLine[0];
				String subtopic = splitedLine[1];
				String docno = splitedLine[2];
				
				int judgment = Integer.parseInt(splitedLine[4]);
				if (judgment == 0) {
					judgment = 1;
				}
				if (!tmpQrels.containsKey(topic)){
					subtopicWeight.put(topic, new HashMap<String,Double>());
					tmpQrels.put(topic, new HashMap<String, Map<String, List<Integer> > >());
				} 
				
				if (!tmpQrels.get(topic).containsKey(docno)){
					
					tmpQrels.get(topic).put(docno, new HashMap<String, List<Integer> > ());
				}
				
				if (!tmpQrels.get(topic).get(docno).containsKey(subtopic)){
					subtopicWeight.get(topic).put(subtopic, 1d);
					tmpQrels.get(topic).get(docno).put(subtopic, new ArrayList<Integer>());
				}
				
				tmpQrels.get(topic).get(docno).get(subtopic).add(judgment);
				
			}
			
		} catch (FileNotFoundException e) {
			
			e.printStackTrace();
		} catch (IOException e) {
			
			e.printStackTrace();
		}
		
		for (Entry<String, Map<String, Map<String, List<Integer> > >> entryTopic : tmpQrels.entrySet()) {
			
			
			String topic = entryTopic.getKey();
			
			if (!qrels.containsKey(entryTopic.getKey())){
				qrels.put(entryTopic.getKey(), new HashMap<String, Map<String, Double > >());
			} 
			
			for (Entry<String, Map<String, List<Integer>>> entryDocno: entryTopic.getValue().entrySet()){
				String docno = entryDocno.getKey();
				if (!qrels.get(topic).containsKey(docno)){
					qrels.get(topic).put(docno, new HashMap<String, Double >());
				}
				
				for(Entry<String, List<Integer>> entrySubtopic: entryDocno.getValue().entrySet()){
					
					String subtopic = entrySubtopic.getKey();	
					Collections.sort(entrySubtopic.getValue());
					Collections.reverse(entrySubtopic.getValue()); 
					List<Integer> rels = entrySubtopic.getValue();
					double rel = 0;
					double log2 = Math.log(2);
					
					for (int i = 0; i < rels.size(); i++) {
						rel += rels.get(i)/(Math.log(i+2)/log2);;
					}
					
					qrels.get(topic).get(docno).put(subtopic, rel);
				}
				
			}
			
		}
		
		// Normalize subtopic weight
		for(Entry<String, Map<String,Double>> entry: subtopicWeight.entrySet()){
			double maxWeight = getMaxWeight(entry.getKey());
			for (Entry<String,Double> subtopic: entry.getValue().entrySet()) {
				subtopic.setValue(subtopic.getValue()/maxWeight);
			}
		}
	}
	
	private double getMaxWeight(String topic) {
		double maxWeight = 0;
		for (Entry<String,Double> subtopic: subtopicWeight.get(topic).entrySet()) {
			maxWeight += subtopic.getValue();
		}
		return maxWeight;
	}

	public double getCubeTest(int iteration, String topic, String[][] docnos){
		currentGainHeight = new HashMap<String, Double>();
		subtopicCover = new HashMap<String, Integer>();
		double score = 0;
		int maxIterations = 0;
		for (int i = 0; i < iteration; i++) {
			if (docnos[i] == null) {
				break;
			}
			maxIterations++;
			for (int j = 0; j < docnos[i].length; j++) {
				if (docnos[i][j] == null) {
					continue;
				}
				double gain = getDocGain(topic,docnos[i][j]);
				score += gain;
			}
			
		}
		maxIterations = Math.min(maxIterations, iteration);
		double ct = score/MAX_HEIGHT;
		double ctSpeed = ct/(maxIterations);
		return ctSpeed;
	}
	
	private double getDocGain(String topic, String docno){
		double docGain = 0;
		for (Entry<String,Double> subtopicEntry : subtopicWeight.get(topic).entrySet()) {
			String subtopic = subtopicEntry.getKey();
			
			if (!currentGainHeight.containsKey(subtopic)){
				currentGainHeight.put(subtopic, 0d);
				subtopicCover.put(subtopic, 0);
				
			}
			
			if (qrels.get(topic).containsKey(docno)){
				if (!qrels.get(topic).get(docno).containsKey(subtopic)){
					continue;
				}

			} else {
				continue;
			}
			
			double area = subtopicEntry.getValue();
			int nrel = subtopicCover.get(subtopic);
			double hightKeepfilling = getHightKeepFilling(topic,docno,subtopic,nrel+1);
			docGain += area*hightKeepfilling;
			
			subtopicCover.put(subtopic, nrel+1);
		}
		return docGain;
	}
	
	private double getHightKeepFilling(String topic, String docno, String subtopic, int nrel){
		double rel = 0;
		
		if (qrels.get(topic).containsKey(docno)) {
			rel = qrels.get(topic).get(docno).getOrDefault(subtopic, 0d);
			if (rel == 0){
				return 0;
			}
		}
		
		double currentGain = currentGainHeight.get(subtopic);
		
		double gain = getHightDiscount(nrel)*rel;
		
		if (currentGain + gain > MAX_HEIGHT){
			gain = MAX_HEIGHT - currentGain;
		}
		
		currentGainHeight.put(subtopic, currentGain+gain);
		return gain;
	
	}
	
	private double getHightDiscount(int nrel) {
		return Math.pow(gamma, nrel);
	}

	public double getAverageCubeTest(int iteration, String topic, String[][] docnos){
		
		currentGainHeight = new HashMap<String, Double>();
		subtopicCover = new HashMap<String, Integer>();
		double score = 0;
		double ctAccu = 0;
		int time = 0;
		for (int i = 0; i < iteration; i++) {
			if (i > docnos.length){
				break;
			}
			if (docnos[i] == null) {
				for (int j = 0; j < 5; j++) {
					//double act = score/MAX_HEIGHT;
					//System.out.println(docnos[i][j]);
					//ctAccu += act/(i+1);
					//time++;
				}
				continue;
			}
			for (int j = 0; j < docnos[i].length; j++) {
				if (docnos[i][j] == null){
					//double act = score/MAX_HEIGHT;
					//System.out.println(docnos[i][j]);
					//ctAccu += act/(i+1);
					//time++;
					continue;
				}
				double gain = getDocGain(topic,docnos[i][j]);
				score += gain;
				double act = score/MAX_HEIGHT;
				//System.out.println(docnos[i][j]);
				ctAccu += act/(i+1);
				time++;
			}
			
		}
		//System.out.println();
		return ctAccu/time;
	}
	
	public double getGain(int iteration, String topic, String[][] docnos){
		
		currentGainHeight = new HashMap<String, Double>();
		subtopicCover = new HashMap<String, Integer>();
		double score = 0;
		double ctAccu = 0;
		int time = 0;
		for (int i = 0; i < iteration; i++) {
			if (i > docnos.length){
				break;
			}
			if (docnos[i] == null) {
				break;
			}
			for (int j = 0; j < docnos[i].length; j++) {
				if (docnos[i][j] == null){
					continue;
				}
				double gain = getDocGain(topic,docnos[i][j]);
				score += gain;
				double act = score/MAX_HEIGHT;
				//System.out.println(docnos[i][j]);
				ctAccu += act/(i+1);
				time++;
			}
			
		}
		//System.out.println();
		return score;
	}
}